internal/clients/aws.go

```diff
@@ -7,61 +7,165 @@ package clients
+		}
+		ps.Configuration[keyAssumeRoleWithWebIdentity] = pc.Spec.Credentials.WebIdentity.RoleARN
+		ps.Configuration[keyWebIdentityTokenFile] = defaultIdentityTokenFile
+	case authKeyUpbound:
+		if pc.Spec.Credentials.Upbound == nil || pc.Spec.Credentials.Upbound.WebIdentity == nil {
+			return terraform.Setup{}, errors.New(`spec.credentials.upbound.webIdentity of ProviderConfig cannot be nil when the credential source is "Upbound"`)
+		}
+		ps.Configuration[keyAssumeRoleWithWebIdentity] = pc.Spec.Credentials.Upbound.WebIdentity.RoleARN
+		ps.Configuration[keyWebIdentityTokenFile] = upboundProviderIdentityTokenFile
+	case authKeySecret:
+		data, err := resource.CommonCredentialExtractor(ctx, s, c, pc.Spec.Credentials.CommonCredentialSelectors)
 		if err != nil {
-			return terraform.Setup{}, errors.Wrap(err, "cannot get AWS config")
+			return terraform.Setup{}, errors.Wrap(err, "cannot get credentials")
 		}
-		if cfg.Region == "" && mg.GetObjectKind().GroupVersionKind().Group == "iam.aws.upbound.io" {
-			cfg.Region = "us-east-1"
+		cfg, err = UseProviderSecret(ctx, data, DefaultSection, cfg.Region)
+		if err != nil {
+			return terraform.Setup{}, errors.Wrap(err, errAWSConfig)
 		}
 		creds, err := cfg.Credentials.Retrieve(ctx)
 		if err != nil {
 			return terraform.Setup{}, errors.Wrap(err, "failed to retrieve aws credentials from aws config")
 		}
-		identity, err := GlobalCallerIdentityCache.GetCallerIdentity(ctx, *cfg, creds)
-		if err != nil {
-			return terraform.Setup{}, errors.Wrap(err, "cannot get the caller identity")
+		ps.Configuration = map[string]any{
+			keyRegion:          cfg.Region,
+			keyAccessKeyID:     creds.AccessKeyID,
+			keySecretAccessKey: creds.SecretAccessKey,
+			keySessionToken:    creds.SessionToken,
 		}
+	}
 
-		ps := terraform.Setup{
-			Version: version,
-			Requirement: terraform.ProviderRequirement{
-				Source:  providerSource,
-				Version: providerVersion,
-			},
-			Configuration: map[string]any{
-				keyRegion:          cfg.Region,
-				keyAccessKeyID:     creds.AccessKeyID,
-				keySecretAccessKey: creds.SecretAccessKey,
-				keySessionToken:    creds.SessionToken,
-			},
-			// Account ID is not part of provider configuration schema, so it
-			// needs to be given separately.
-			ClientMetadata: map[string]string{
-				keyAccountId: *identity.Account,
-			},
-		}
-		return ps, err
+	if len(pc.Spec.AssumeRoleChain) > 1 || pc.Spec.Endpoint != nil {
+		return terraform.Setup{}, errors.New("shared scheduler cannot be used because the length of assume role chain array " +
+			"is more than 1 or endpoint configuration is not nil")
+	}
+
+	if len(pc.Spec.AssumeRoleChain) != 0 {
+		ps.Configuration[keyAssumeRole] = pc.Spec.AssumeRoleChain[0].RoleARN
+		ps.Configuration[keyTags] = pc.Spec.AssumeRoleChain[0].Tags
+		ps.Configuration[keyTransitiveTagKeys] = pc.Spec.AssumeRoleChain[0].TransitiveTagKeys
```

Comment:
We also need to incorporate the `pc.Spec.AssumeRoleChain[0].ExternalID`:
<img width="561" alt="image" src="https://user-images.githubusercontent.com/9376684/227198980-9512b8f9-56d1-44b4-b2e9-68ef06b3730e.png">